dv/cosim/spike_cosim.cc

```diff
@@ -0,0 +1,516 @@
+}
+
+bool SpikeCosim::backdoor_read_mem(uint32_t addr, size_t len,
+                                   uint8_t *data_out) {
+  return bus.load(addr, len, data_out);
+}
+
+bool SpikeCosim::step(uint32_t write_reg, uint32_t write_reg_data, uint32_t pc,
+                      bool sync_trap) {
+  assert(write_reg < 32);
+
+  if (pc_is_mret(pc)) {
+    nmi_mode = false;
+  }
+
+  uint32_t initial_pc = (processor->get_state()->pc & 0xffffffff);
+  bool initial_pc_match = initial_pc == pc;
+
+  // Execute the next instruction
+  processor->step(1);
+
+  if (processor->get_state()->last_inst_pc == PC_INVALID) {
+    if (processor->get_state()->mcause->read() & 0x80000000) {
+      // Interrupt occurred, step again to execute first instruction of
+      // interrupt
+      processor->step(1);
+      // TODO: Deal with exception on first instruction of interrupt
+      assert(processor->get_state()->last_inst_pc != PC_INVALID);
+    } else {
+      // Otherwise a synchronous trap has occurred, check the DUT reported a
+      // synchronous trap at the same point
+      if (!sync_trap) {
+        std::stringstream err_str;
+        err_str << "Synchronous trap was expected at ISS PC: " << std::hex
+                << processor->get_state()->pc
+                << " but DUT didn't report one at PC " << pc;
+        errors.emplace_back(err_str.str());
+
+        return false;
+      }
+
+      if (!initial_pc_match) {
+        std::stringstream err_str;
+        err_str << "PC mismatch at synchronous trap, DUT: " << std::hex << pc
+                << " expected: " << std::hex << initial_pc;
+        errors.emplace_back(err_str.str());
+
+        return false;
+      }
+
+      if (write_reg != 0) {
+        std::stringstream err_str;
+        err_str << "Synchronous trap occurred at PC: " << std::hex << pc
+                << "but DUT wrote to register: x" << std::dec << write_reg;
+        errors.emplace_back(err_str.str());
+
+        return false;
+      }
+
+      return errors.size() == 0;
+    }
+  }
+
+  // Check PC of executed instruction matches the expected PC
+  // TODO: Confirm details of why spike sign extends PC, something to do with
+  // 32-bit address as 64-bit address must be sign extended?
+  if ((processor->get_state()->last_inst_pc & 0xffffffff) != pc) {
+    std::stringstream err_str;
+    err_str << "PC mismatch, DUT: " << std::hex << pc
+            << " expected: " << std::hex
+            << processor->get_state()->last_inst_pc;
+    errors.emplace_back(err_str.str());
+
+    return false;
+  }
+
+  // Check register writes from executed instruction match what is expected
+  auto &reg_changes = processor->get_state()->log_reg_write;
+
+  bool gpr_write_seen = false;
+
+  for (auto reg_change : reg_changes) {
+    // reg_change.first provides register type in bottom 4 bits, then register
+    // index above that
+
+    // Ignore writes to x0
+    if (reg_change.first == 0)
+      continue;
+
+    if ((reg_change.first & 0xf) == 0) {
```

Comment:
I'd consider moving the `on_gpr_write` / `on_csr_write` functionality into separate helper functions. It might make the (rather mammoth) `step()` method a bit easier to understand.